parser/transaction.go

```diff
@@ -397,76 +381,193 @@ func (tx *Transaction) ParseFromSlice(data []byte) ([]byte, error) {
-			return nil, errors.New("could not read bindingSig")
+	var spendCount, outputCount int
+	if !s.ReadCompactSize(&spendCount) {
+		return nil, errors.New("could not read nShieldedSpend")
+	}
+	tx.shieldedSpends = make([]*spend, spendCount)
+	for i := 0; i < spendCount; i++ {
+		newSpend := &spend{}
+		s, err = newSpend.ParseFromSlice([]byte(s), tx.version)
+		if err != nil {
+			return nil, errors.Wrap(err, "while parsing shielded Spend")
+		}
+		tx.shieldedSpends[i] = newSpend
+	}
+	if !s.ReadCompactSize(&outputCount) {
+		return nil, errors.New("could not read nShieldedOutput")
+	}
+	tx.shieldedOutputs = make([]*output, outputCount)
+	for i := 0; i < outputCount; i++ {
+		newOutput := &output{}
+		s, err = newOutput.ParseFromSlice([]byte(s), tx.version)
+		if err != nil {
+			return nil, errors.Wrap(err, "while parsing shielded Output")
+		}
+		tx.shieldedOutputs[i] = newOutput
+	}
+	if spendCount+outputCount > 0 && !s.ReadInt64(&tx.valueBalanceSapling) {
+		return nil, errors.New("could not read valueBalance")
+	}
+	if spendCount > 0 && !s.Skip(32) {
+		return nil, errors.New("could not skip anchorSapling")
+	}
+	if !s.Skip(192 * spendCount) {
+		return nil, errors.New("could not skip vSpendProofsSapling")
+	}
+	if !s.Skip(64 * spendCount) {
+		return nil, errors.New("could not skip vSpendAuthSigsSapling")
+	}
+	if !s.Skip(192 * outputCount) {
+		return nil, errors.New("could not skip vOutputProofsSapling")
+	}
+	if spendCount+outputCount > 0 && !s.ReadBytes(&tx.bindingSigSapling, 64) {
+		return nil, errors.New("could not read bindingSigSapling")
+	}
+	var actionsCount, proofsCount int
+	if !s.ReadCompactSize(&actionsCount) {
+		return nil, errors.New("could not read nActionsOrchard")
+	}
+	if !s.Skip(820 * actionsCount) {
+		return nil, errors.New("could not skip vActionsOrchard")
+	}
+	if actionsCount > 0 {
+		if !s.Skip(1) {
+			return nil, errors.New("could not skip flagsOrchard")
+		}
+		if !s.Skip(8) {
+			return nil, errors.New("could not skip valueBalanceOrchard")
+		}
+		if !s.Skip(32) {
+			return nil, errors.New("could not skip anchorOrchard")
+		}
+		if !s.ReadCompactSize(&proofsCount) {
+			return nil, errors.New("could not read sizeProofsOrchard")
+		}
+		if !s.Skip(proofsCount) {
+			return nil, errors.New("could not skip proofsOrchard")
+		}
+		if !s.Skip(64 * actionsCount) {
+			return nil, errors.New("could not skip vSpendAuthSigsOrchard")
+		}
+		if !s.Skip(64) {
+			return nil, errors.New("could not skip bindingSigOrchard")
 		}
 	}
+	return s, nil
+}
+
+// ParseFromSlice deserializes a single transaction from the given data.
+func (tx *Transaction) ParseFromSlice(data []byte) ([]byte, error) {
+	s := bytestring.String(data)
+
+	// declare here to prevent shadowing problems in cryptobyte assignments
+	var err error
+
+	var header uint32
+	if !s.ReadUint32(&header) {
+		return nil, errors.New("could not read header")
+	}
+
+	tx.fOverwintered = (header >> 31) == 1
+	tx.version = header & 0x7FFFFFFF
```

Comment:
Maybe add a check here to ensure that the overwintered flag is properly set if the version is `>= 3`. If not, then we can't trust the meaning of `version`: 

> Version constraints apply to the effectiveVersion, which is equal to min(2, version) when fOverwintered = 0 and to version otherwise.

We want to be able to rely upon the value of `tx.version` for the rest of our parsing, but if `fOverwintered == 0` then we can't do so. So either all the `tx.version >= 3` constraints also need an `fOverwintered == 1` check, or we need to raise an error here.

Reply:
> ... but if `fOverwintered == 0` then we can't do so.

The spec says that if `fOverwintered` is false, then we treat a parsed version that's greater than 2 as if it's 2. So in theory, we could parse version 10, but if `fOverwintered` is false, then we act as if version is 2. (If version is 1, then we leave it as a 1).

The version 5 part of the spec also says that `fOverwintered` must be set. (But it does not say this for earlier versions.)

I'll push a change that I think follows the spec, let me know what you think.